app/featuresmodel.cpp

```diff
@@ -25,31 +35,77 @@ void FeaturesModel::populate()
 {
   if ( mLayer )
   {
+    mFetchingResults = true;
+    emit fetchingResultsChanged();
     beginResetModel();
     mFeatures.clear();
+    endResetModel();
 
     QgsFeatureRequest req;
     setupFeatureRequest( req );
 
-    QgsFeatureIterator it = mLayer->getFeatures( req );
-    QgsFeature f;
+    int searchId = mNextSearchId.fetchAndAddOrdered( 1 );
+    QgsVectorLayer *layer = mLayer->clone();
```

Comment:
cloning a whole layer for this may be a bit overkill. You could use `QgsVectorLayerFeatureSource` just like it is done in QgsVectorLayerRenderer: in the main thread one creates instance of QgsVectorLayerFeatureSource from layer, and then in the worker thread one just calls source.getFeatures(). This should be much less work (and no QObjects, so no need to do explicit moveToThread() calls).

Reply:
done!